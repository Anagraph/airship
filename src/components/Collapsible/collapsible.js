import React, { Component, Children } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import PlusIcon from '../Icons/plus';
import CloseIcon from '../Icons/close';
import MinusIcon from '../Icons/minus';

/*
  <Collapsible>
    <Collapsible.Header>Header</Collapsible.Header>
    <Collapsible.Content>Content</Collapsible.Content>
  </Collapsible>
*/

const Header = styled.div`
  display: flex;

  button {
    border: 0;
    background: transparent;
    box-shadown: none;
    margin-left: auto;
    display: flex;
    align-items: center;
    outline: none;
  }
`;
Header.displayName = 'Collapsible.Header';

const Content = ({ children, ...props }) => {
  return children;
};
Content.displayName = 'Collapsible.Content';

class Collapsible extends Component {
  static Header = Header;
  static Content = Content;

  state = {
    open: !!this.props.open
  };

  componentWillReceiveProps(nextProps) {
    if (this.state.open !== nextProps.open) {
      this.setState({ open: nextProps.open });
    }
  }

  toggle = () => {
    const { onChange } = this.props;

    this.setState(
      (state) => {
        return {
          ...state,
          open: !state.open
        };
      },
      () => {
        onChange && onChange(this.state);
      }
    );
  };

  render() {
    const { open } = this.state;
    const { children } = this.props;

    return (
      <div>
        {Children.map(children, (child) => {
          if (child.type.displayName === 'Collapsible.Header') {
            return (
              <Collapsible.Header>
                <div>{child.props.children}</div>
                <button
                  onClick={(e) => {
                    this.toggle();
                  }}
                >
                  {open ? <MinusIcon /> : <PlusIcon />}
                </button>
              </Collapsible.Header>
            );
          } else if (child.type.displayName === 'Collapsible.Content' && open) {
            return child;
          }
        })}
      </div>
    );
  }
}

Collapsible.defaultProps = {
  open: true
};

Collapsible.propTypes = {
  open: PropTypes.bool,
  onChange: PropTypes.func
};

export default Collapsible;
