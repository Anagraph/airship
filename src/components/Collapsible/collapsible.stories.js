import React from 'react';
import { storiesOf } from '@storybook/react';
import Collapsible from '../Collapsible/collapsible';
import Badge from '../Badge/badge';
import Text from '../Typography/text';
import Subheader from '../Typography/subheader';

storiesOf('Collapsible', module)
  .add('Default', () => (
    <div style={{ maxWidth: '200px' }}>
      <Collapsible>
        <Collapsible.Header>
          <Subheader>Global index</Subheader>
        </Collapsible.Header>
        <Collapsible.Content>
          <Badge color="#B4E0FA">Store 1</Badge>
          <Text as="div">Some fanzy text here</Text>
        </Collapsible.Content>
      </Collapsible>
    </div>
  ));
