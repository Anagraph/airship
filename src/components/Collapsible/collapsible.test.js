import React from 'react';
import Collapsible from './collapsible';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(
      <Collapsible>
        <Collapsible.Header>Header</Collapsible.Header>
        <Collapsible.Content>Content</Collapsible.Content>
      </Collapsible>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders open', () => {
    const component = shallow(
      <Collapsible open={true}>
        <Collapsible.Header>Header</Collapsible.Header>
        <Collapsible.Content>Content</Collapsible.Content>
      </Collapsible>
    );
    expect(component.find(Collapsible.Content).length).toBe(1);
  });

  it('callback', () => {
    const spy = jest.fn();
    const component = mount(
      <Collapsible onChange={spy}>
        <Collapsible.Header>Header</Collapsible.Header>
        <Collapsible.Content>Content</Collapsible.Content>
      </Collapsible>
    );

    component.instance().toggle();

    expect(spy).toHaveBeenCalled();
  });
});
