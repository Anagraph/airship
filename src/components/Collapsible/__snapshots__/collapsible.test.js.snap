// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders without crashing 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 12px;
}

.c0 button {
  border: 0;
  background: transparent;
  box-shadown: none;
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: none;
}

<div>
  <div
    className="c0"
  >
    <div>
      Header
    </div>
    <button
      onClick={[Function]}
    >
      <svg
        fill="#1785fb"
        height={8}
        viewBox="0 0 16 8"
        width={16}
      >
        <path
          d="M7.69011573,0.876287597 L8.30988427,0.876287597 L0.780478575,7.8762816 C0.584885024,8.05812247 0.287585208,8.03706374 0.116440851,7.8292456 C-0.0547035068,7.62142746 -0.034883519,7.30554642 0.160710032,7.12370555 L7.69011573,0.123711547 C7.86753958,-0.0412371822 8.13246042,-0.0412371822 8.30988427,0.123711547 L15.83929,7.12370555 C16.0348835,7.30554642 16.0547035,7.62142746 15.8835591,7.8292456 C15.7124148,8.03706374 15.415115,8.05812247 15.2195214,7.8762816 L7.69011573,0.876287597 Z"
        />
      </svg>
    </button>
  </div>
  Content
</div>
`;
